File: context.py
from datetime import datetime, timedelta
from typing import Any, Callable, Dict, List, Optional

from cache import CacheInstance
import data


class Context:
    def __init__(
        self,
        config: Dict,
        global_cache: Optional[Dict],
        client_cache: Optional[Dict],
        shell: str,
        width: int,
        exit_status: int,
        pipe_status: List[int],
        jobs_running: int,
        jobs_sleeping: int,
        current_dir: str,
        envs: Dict[str, str],
    ):
        self.config = config
        self.inst_cache: Dict[str, Any] = dict()
        self.timed_caches = bool(global_cache and client_cache)
        self.global_cache = global_cache or dict()
        self.client_cache = client_cache or dict()

        self.shell = shell
        self.width = width
        self.exit_status = exit_status
        self.pipe_status = pipe_status
        self.jobs_running = jobs_running
        self.jobs_sleeping = jobs_sleeping
        self.current_dir = current_dir
        self.environment_vars = envs

    def with_inst_cache(
        self,
        func: Callable[[Any, Dict], Any],
    ) -> Any:
        key = str(func)
        if key in self.inst_cache:
            return self.inst_cache[key]

        value: str = func(self, self.environment_vars)

        self.inst_cache[key] = value

        return value

    def with_cache(
        self,
        cache: Dict[str, CacheInstance],
        func: Callable[[Any, Dict], Any],
        timeout: timedelta,
    ) -> Any:
        if not self.timed_caches:
            return self.with_inst_cache(func)

        key = str(func)
        if key in cache and not cache[key].is_expired():
            return cache[key].value

        value: str = func(self, self.environment_vars)

        cache[key] = CacheInstance(value, datetime.now() + timeout)

        return value

    def hostname(self) -> str:
        return self.with_cache(self.global_cache, data.hostname, timedelta(minutes=5))

    def username(self) -> str:
        return self.with_cache(self.client_cache, data.username, timedelta(minutes=10))

    def work_dir(self) -> str:
        return self.with_inst_cache(data.work_dir)

    def shell_level(self) -> int:
        return self.with_inst_cache(data.shell_level)

File: src/context.rs
use std::collections::HashMap;
use std::fmt;
use std::sync::RwLock;
use std::vec::Vec;

#[derive(clap::ValueEnum, Clone, Copy, Debug, PartialEq, Eq, rune::Any)]
pub enum Shell {
    Bash,
    Unknown,
}

impl Shell {
    pub fn from_str(input: &str) -> Self {
        let input = match input.rsplit_once('/') {
            Some((_, end)) => end,
            None => input,
        };

        match input.to_lowercase().as_ref() {
            "bash" => Self::Bash,
            _ => Self::Unknown,
        }
    }
}

impl fmt::Display for Shell {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Shell::Bash => write!(f, "bash"),
            Shell::Unknown => write!(f, "unknown"),
        }
    }
}

#[derive(Debug, rune::Any)]
pub struct Context {
    global: RwLock<HashMap<&'static str, String>>,
    client: RwLock<HashMap<&'static str, String>>,
    instance: RwLock<HashMap<&'static str, String>>,
    shell: Shell,
    width: u16,
    exit_status: i32,
    pipe_status: Vec<i32>,
    jobs_running: u16,
    jobs_sleeping: u16,
    current_dir: String,
    pub envs: HashMap<String, String>,
}

macro_rules! with_cache {
    ($cache:expr, $key:expr, $func:expr) => {
        let map = $cache.read().unwrap();
        if map.contains_key($key) {
            return map.get($key).unwrap().clone();
        }
        drop(map);

        let value = $func;

        let mut map = $cache.write().unwrap();
        map.insert($key, value.clone());

        return value;
    };
}

impl Context {
    pub fn new(
        shell: Shell,
        width: u16,
        exit_status: i32,
        pipe_status: Vec<i32>,
        jobs_running: u16,
        jobs_sleeping: u16,
        current_dir: String,
        envs: HashMap<String, String>,
    ) -> Self {
        Context {
            // TODO: actual caching.
            global: RwLock::new(HashMap::new()),
            client: RwLock::new(HashMap::new()),
            instance: RwLock::new(HashMap::new()),
            shell,
            width,
            exit_status,
            pipe_status,
            jobs_running,
            jobs_sleeping,
            current_dir,
            envs,
        }
    }

    pub async fn hostname(&self) -> String {
        with_cache!(
            self.global,
            "hostname",
            crate::data::hostname::hostname(self).await
        );
    }

    pub fn shell(&self) -> Shell {
        self.shell
    }

    pub async fn username(&self) -> String {
        with_cache!(
            self.global,
            "username",
            crate::data::username::username(self).await
        );
    }
}
